src/core/engines/engine_manager.cpp

```diff
@@ -0,0 +1,179 @@
+#include "engine_manager.hpp"
+
+#include "core/ui/mainui.hpp"
+#include "engines/drums/drum-sampler/drum-sampler.hpp"
+#include "engines/drums/simple-drums/simple-drums.hpp"
+#include "engines/synths/nuke/nuke.hpp"
+#include "services/state.hpp"
+
+namespace otto::engines {
+  namespace {
+    std::map<std::string, std::function<AnyEngine*()>> engineGetters;
+    audio::ProcessBuffer<1> _audiobuf1;
+
+    EngineDispatcher<engines::EngineType::synth> synth;
+    EngineDispatcher<engines::EngineType::drums> drums;
+    EngineDispatcher<engines::EngineType::effect> effect;
+    Tapedeck tapedeck;
+    Mixer mixer;
+    Metronome metronome;
+    InputSelector selector;
+  } // namespace
+
+  void init()
+  {
+    engineGetters = {{"TapeDeck", [&]() { return (AnyEngine*) &tapedeck; }},
+                     {"Mixer", [&]() { return (AnyEngine*) &mixer; }},
+                     {"Synth", [&]() { return (AnyEngine*) &*synth; }},
+                     {"Drums", [&]() { return (AnyEngine*) &*drums; }},
+                     {"Metronome", [&]() { return (AnyEngine*) &metronome; }}};
+
+    register_engine<DrumSampler>();
+    register_engine<SimpleDrumsEngine>();
+    register_engine<NukeSynth>();
+
+    synth.init();
+    drums.init();
+
+    ui::registerKeyHandler(ui::Key::play, [](ui::Key key) {
+      if (tapedeck.state.playing()) {
+        tapedeck.state.stop();
+      } else {
+        tapedeck.state.play();
+      }
+    });
+
+    auto load = [&](nlohmann::json& data) {
+      from_json(data["TapeDeck"], tapedeck);
+      from_json(data["Mixer"], mixer);
+      from_json(data["Synth"], synth);
+      from_json(data["Drums"], drums);
+      from_json(data["Metronome"], metronome);
+    };
+
+    auto save = [&]() {
+      return nlohmann::json({{"TapeDeck", tapedeck},
+                             {"Mixer", mixer},
+                             {"Synth", synth},
+                             {"Drums", drums},
+                             {"Metronome", metronome}});
+    };
+
+    services::state::attach("Engines", load, save);
+  }
+
+  void start()
+  {
+    tapedeck.on_enable();
+    metronome.on_enable();
+    mixer.on_enable();
+    synth.select(std::size_t(0));
+    drums.select(std::size_t(0));
+  }
+
+  void shutdown()
+  {
+    mixer.on_disable();
+    tapedeck.on_disable();
+  }
+
+  audio::ProcessData<2> processAudio(
```

Comment:
while these could be renamed to `process_audio`, all these process functions should be the same. I've also considered making a more generic interface for them, but once again, not for now.
  
also: snake_case_it